gem/impero_utils.py

```diff
@@ -256,6 +256,7 @@ def recurse_block(expr, loop_indices):
                 # Remove expression from the ``refcount`` so it will
                 # not be declared again.
                 del refcount[e]
+        declare[expr].sort(key=lambda t: numbering[t])
```

Comment:
Now you just shove stuff into the list of declarations in any old order and at the end sort with semantically the same as `temporaries.index`, but with an `O(1)` comparator.